### PR TITLE
[3d] Wide lines + polygon edge highlighting

### PR DESCRIPTION
**Pull request by wonder-sk**
From user's point of view, this adds:
- configurable line width (in screen units) for "simple" 3D lines - until now they were always 1px wide
- optional highlighting of edges of 3D polygons

PolygonZ data of buildings (Prague open data) with edge highlighting:

image

This was made possible thanks to @andreasneumann !

Some thoughts about line rendering. Until now we had two completely different approaches:
- "simple" lines - 1px thick, supporting Z coordinates
- "buffered" lines - thickness defined in map units. Not supporting Z coordinates (basically we did buffer operation so all Z coordinates were discarded)

Now this new approach fixes "simple" lines so they can have any thickness. The "simple" style should probably become the default. Moreover I am wondering if it makes sense to keep the "buffered" lines style still around...

For those curious about more details:
- "simple" line rendering used OpenGL's LineStrip primitive before. This is generally not that well supported across platforms and can lead to inconsistent results (with varying degree of wide line support, but generally drawing bad looking lines)
- the PR adds a new Qt3D material for drawing lines. The technique uses LineStripAdjacency primitive and a geometry shader that turns line segments into pairs of triangles. The "adjacency" bit in the primitive name means that for every single line segment we know the previous and the next point in the geometry shader and we can create line caps/joins as needed
- it comes with flat caps at the end of lines and either bevel or miter line joins. It is capable of drawing lines of any thickness. It is also possible to provide a 1xN pixels image with line cross-section which would be used for drawing (instead of just plain color), this is currently not supported in the GUI
- edge highlighting is done by rendering line strips of original polygon data using the line material. This implies that if the input geometry is a wall/roof polygon, it gets outlines as expected. But if the input geometry is already triangulated, the result is that e.g. a rectangular wall also gets a line across the corners as it is split into two triangles. That's why it is much better to work with data that has not been triangulated yet. "Multipatch" shapefile data are triangulated already, on the other hand "polyhedral surface" or "polygonZ" data should not be triangulated.
- the way how edge highlighting is done also means that some edges may not get highlighted. For example, if two solid objects intersect each other, but this intersection is not specified in data, there would be no edges shown. This however does not seem to be common especially with building data. Originally I started with a completely different approach where a scene was first rendered and then a post-processing step in screen space was done on a combination of normal map and depth map - it searched for edges using Sobel filter. This had the advantage of being true edge detector, discovering also edges that were not explicitly specified in original data. But it had various downsides that seemed to outweigh that - for example it was not possible to use arbitrary width of edges, detected edges did not look great (they had alias), it may have needed some tweaking and it required more complex rendering pipeline structure (1. render entities that need edge detection as normal/depth maps, 2. render the whole scene, 3. apply post-processing)

### CODE REVIEW

File: src/3d/CMakeLists.txt [hunk 1]
```diff
@@ -122,6 +126,8 @@ SET(QGIS_3D_HDRS
   symbols/qgsabstract3dsymbol.h
   symbols/qgsline3dsymbol.h
   symbols/qgsline3dsymbol_p.h
+  symbols/qgslinematerial_p.h
+  symbols/qgslinevertexdata_p.h
```
Comment (m-kuhn) [inline src/3d/CMakeLists.txt]: I think private headers shouldn't be installed normally, so no reason to list them here

File: src/3d/symbols/qgslinevertexdata_p.cpp [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+}
+
+
+void QgsLineVertexData::addVerticalLines( const QgsLineString &lineString, float verticalLength )
+{
+  QgsPoint centroid;
+  if ( altBinding == Qgs3DTypes::AltBindCentroid )
+    centroid = lineString.centroid();
+
+  for ( int i = 0; i < lineString.vertexCount(); ++i )
```
Comment (m-kuhn) [inline src/3d/symbols/qgslinevertexdata_p.cpp]: Didn't you write a vertex iterator?

**Comment by nyalldawson** [comment]
Cool feature! 

Are you going to add a unit test for this too? If so, better use a super chunky line width to make it robust

File: src/3d/CMakeLists.txt (hunk 1 above)
Comment (wonder-sk) [reply src/3d/CMakeLists.txt]: There are few cases when private include files may be transitively #included by some public headers so for now it seemed better to install these as well. Once we will declare stable api for the 3d library, we can clean that up (what is public and what is private may still change)...

File: src/3d/symbols/qgslinevertexdata_p.cpp (hunk 2 above)
Comment (wonder-sk) [reply src/3d/symbols/qgslinevertexdata_p.cpp]: Yes, vertex iterator is good when using arbitrary geometries. But when dealing with a particular type of geometry, why not use something with less overhead :-)

**Comment by wonder-sk** [comment]
Rendering tests added!

File: src/3d/symbols/qgslinevertexdata_p.cpp (hunk 2 above)
Comment (m-kuhn) [reply src/3d/symbols/qgslinevertexdata_p.cpp]: I was assuming that iterators have less overhead than repeatedly calling vertexCount() and looking up the item at a given index, but I admit I didn't do any check for ground truth :)

File: src/3d/CMakeLists.txt (hunk 1 above)
Comment (m-kuhn) [reply src/3d/CMakeLists.txt]: > There are few cases when private include files may be transitively #included by some public header

But that's not the case here, is it?

File: src/3d/shaders/lines.frag [hunk 3]
```diff
@@ -0,0 +1,30 @@
+#version 150
```
Comment (m-kuhn) [inline src/3d/shaders/lines.frag]: Should these files have license headers too?

**Comment by m-kuhn** [comment]
Looks really nice :+1:

File: src/3d/shaders/lines.frag (hunk 3 above)
Comment (wonder-sk) [reply src/3d/shaders/lines.frag]: Good question... just checked Qt sources and they did not include any license headers in shaders. I'm quite neutral in this regard. Two small advantages when they are left out: 1. shader compilation is done in run-time, so less amount of text to compile, 2. on shader compilation error the shader is printed along with the error, so extra license headers can make the output unnecessarily verbose.

File: src/3d/CMakeLists.txt (hunk 1 above)
Comment (wonder-sk) [reply src/3d/CMakeLists.txt]: Nope, just being consistent by adding them so we do not wonder later why some private headers are installed and some aren't :)

File: src/3d/shaders/lines.frag (hunk 3 above)
Comment (m-kuhn) [reply src/3d/shaders/lines.frag]: Also no strong opinion here, but these are 2 good reasons not to include them.

**Pull request merged**